aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/ModelTreeHelper.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.datastore.storage.sqlite;
+
+import android.database.Cursor;
+
+import com.amplifyframework.core.model.Model;
+import com.amplifyframework.core.model.ModelAssociation;
+import com.amplifyframework.core.model.ModelSchema;
+import com.amplifyframework.core.model.ModelSchemaRegistry;
+import com.amplifyframework.core.model.query.QueryOptions;
+import com.amplifyframework.core.model.query.Where;
+import com.amplifyframework.core.model.query.predicate.QueryField;
+import com.amplifyframework.core.model.query.predicate.QueryPredicate;
+import com.amplifyframework.core.model.query.predicate.QueryPredicateOperation;
+import com.amplifyframework.core.model.query.predicate.QueryPredicates;
+import com.amplifyframework.datastore.DataStoreException;
+import com.amplifyframework.datastore.storage.sqlite.adapter.SQLiteTable;
+import com.amplifyframework.util.Empty;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utility class to help traverse a tree of models by relationship.
+ */
+final class ModelTreeHelper {
+    private final ModelSchemaRegistry registry;
+    private final SQLiteStorageAdapter storage;
+
+    /**
+     * Constructs a model family tree traversing utility.
+     * @param registry model registry to search schema from
+     * @param storage SQLite storage engine
+     */
+    ModelTreeHelper(ModelSchemaRegistry registry,
+                    SQLiteStorageAdapter storage) {
+        this.registry = registry;
+        this.storage = storage;
+    }
+
+    /**
+     * Returns a map of descendants of a set of models (of same type).
+     * A model is a child of its parent if it uses its parent's ID as foreign key.
+     * @param root Collection of models to query its descendants of.
+     * @return Map of descendants keyed by model schema. The value contains a set of
+     *          descendants' IDs for that model type.
+     */
+    <T extends Model> Map<ModelSchema, Set<String>> descendantsOf(Collection<T> root) {
```

Comment:
This type still kind of looks like a data structure in-and-of itself. You could make a `class Descendent`, and then return a `List<Descendent>`, here.
```
    Map<ModelSchema, Set<String>>
```

Any Map other than `Map<PrimitiveA, PrimitiveB>` is usually waiting to be made into a complex type of its own.

Reply:
I still do need to be able to search for the set of IDs by model, which warrants the use of a map. Only way to simplify this further would be to use `Map<PrimitiveA, NewType>` where `NewType` wraps a set, but that would be an overkill. I think a "Map of Collections" is an intuitive enough data structure that it won't need another level of simplification.